[GLib] Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebKitWebViewSessionState.cpp

aperezdc opened the pull request:
#### 2657decc298438aaa7a7e281977441deca4b402b
<pre>
[GLib] Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebKitWebViewSessionState.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=303231">https://bugs.webkit.org/show_bug.cgi?id=303231</a>

Reviewed by Carlos Garcia Campos.

* Source/WebKit/UIProcess/API/glib/WebKitWebViewSessionState.cpp:
(decodeFrameState): Create the String/AtomString instances first, and
check their whether they are not empty instead of using strlen().
(decodeSessionState): Use ASCIILiteral instead of plain C strings, replace
bare array with std::array.

Canonical link: <a href="https://commits.webkit.org/303647@main">https://commits.webkit.org/303647@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/371659aaa291f863cac2607e4341d07b4a7a28e6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133156 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5657 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44273 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140709 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85196 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6166 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5520 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101842 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69260 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136103 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4368 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119338 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82638 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4253 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1821 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113330 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143358 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5327 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38033 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110219 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5409 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4575 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110401 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27976 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4120 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115595 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59042 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5382 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33949 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5227 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68834 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5471 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5338 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebViewSessionState.cpp:
```diff
@@ -407,31 +405,31 @@ static inline void decodeBackForwardListItemState(GVariantIter* backForwardListS
-        sessionStateTypeString = sessionStateTypeStringVersions[i];
-        variant = g_variant_new_from_bytes(G_VARIANT_TYPE(sessionStateTypeString), data, FALSE);
-        if (g_variant_is_normal_form(variant.get()))
+    const ASCIILiteral* sessionStateTypeString = nullptr;
```
Do you need to use a pointer?

aperezdc replied:
Turns out that `ASCIILiteral` doesn't have a copy-constructor nor an assignment operator. And for [Reasons™](https://www.fluentcpp.com/2018/10/05/pros-cons-optional-references/) using `std::optional<T&>` is not a thing... oh well, I suppose one could use [std::reference_wrapper](https://en.cppreference.com/w/cpp/utility/functional/reference_wrapper.html) 🤔

aperezdc replied:
Well, I just realized that we can remove the `sessionStateTypeString` variable, and use `g_variant_get_type_string()` down below. Nicer.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebViewSessionState.cpp:
```diff
@@ -407,31 +405,31 @@ static inline void decodeBackForwardListItemState(GVariantIter* backForwardListS
+    for (const auto& typeString : sessionStateTypeStringVersions) {
+        GRefPtr<GVariant> maybeVariant = g_variant_new_from_bytes(G_VARIANT_TYPE(typeString), data, FALSE);
+        if (g_variant_is_normal_form(maybeVariant.get())) {
+            sessionStateTypeString = &typeString;
```
Maybe it would be clearer if we initialize the index and then we use sessionStateTypeStringVersions [index]

aperezdc committed:
[GLib] Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebKitWebViewSessionState.cpp
https://bugs.webkit.org/show_bug.cgi?id=303231

Reviewed by Carlos Garcia Campos.

* Source/WebKit/UIProcess/API/glib/WebKitWebViewSessionState.cpp:
(decodeFrameState): Create the String/AtomString instances first, and
check their whether they are not empty instead of using strlen().
(decodeSessionState): Use ASCIILiteral instead of plain C strings, replace
bare array with std::array.

Canonical link: https://commits.webkit.org/303647@main
